lib/util.js

```diff
@@ -152,6 +152,51 @@ async function streamToBuffer(method, url, status, stream, totalSize) {
     });
 }
 
+/**
+ * Converts the content of a readable Stream into a Buffer
+ * instance. WARNING: this method will read the entirety of
+ * the stream into memory; use with extreme caution.
+ *
+ * @param {String} method HTTP method
+ * @param {String} url URL requested
+ * @param {Number} status HTTP status code of response
+ * @param {ReadableStream} stream Stream to convert.
+ * @param {number} totalSize Total size of the stream.
+ * @returns {Promise<TransferMemoryBlock>} Contents of the stream in a
+ *  pooled buffer memory block.
+ */
+async function streamToPooledBuffer(method, url, status, stream, totalSize, bufferBlockMemoryAllocator) {
+    if (!bufferBlockMemoryAllocator) {
+        return streamToBuffer(method, url, status, stream, totalSize);
+    }
+
+    return bufferBlockMemoryAllocator.obtainBuffer(totalSize).then((bufferMemoryBlock) => {
+        return new Promise((resolve, reject) => {
+            let bufferOffset = 0;
+            stream.once('error', (err) => {
+                reject(new HttpStreamError(method, url, status, err));
+            });
+
+            stream.once('end', () => {
+                if (bufferOffset !== totalSize) {
+                    reject(new HttpStreamError(method, url, status, `Unexpected number of bytes read from stream. Expected ${totalSize} but got ${bufferOffset}.`));
+                    return;
+                }
+                resolve(bufferMemoryBlock);
+            });
+
+            stream.on('data', (chunk) => {
+                if(bufferOffset >= totalSize){
+                    reject(new HttpStreamError(method, url, status, `Unexpected number of bytes written to buffer from stream. Too many bytes: Expected ${totalSize} but got ${bufferOffset}.`));
+                }
+
+                chunk.copy(bufferMemoryBlock.buffer, bufferOffset, 0);
+                bufferOffset += chunk.length;
+            });
+        });
+    });
+}
```

Comment:
Same here as the RandomFileAccess class, pass in the buffer so the memory allocation and release are done in the same code block.

Reply:
As explained in previous comment, the buffer is not passed in directly to stay encapsulated and under control of the memory allocator (since the memory allocator handles the underlying buffer slicing and allocation). The memory allocator instance  is passed-in instead, to give out buffer slices as needed, and to be notified by the client code when it's done with the buffers so they can be released back to available memory allocatable by the memory allocator.

Reply:
It's not encapsulated though, from transfer:

```
const bufferMemoryBlock = await streamToPooledBuffer(HTTP.METHOD.GET, transferPart.source.url, response.status, response.body, contentLength, this.randomFileAccess.bufferMemoryAllocator);
await this.randomFileAccess.write(targetUrl, contentRange, bufferMemoryBlock.buffer, totalSize);
this.randomFileAccess.releaseFileBufferMemory(bufferMemoryBlock);
```

There is wrapping which requires going a layer deeper to understand how memory management is handled. It would be better if transfer requests the memory, uses it to transfer and then hands it back to the allocator.

Reply:
If it's ok to change the APIs here, I can make the change. If it's not, we'll have to live with that though.